test/integration/src/test/java/io/pravega/test/integration/RestoreBackUpDataRecoveryTest.java

```diff
@@ -0,0 +1,648 @@
+        readAllEvents(STREAM2, clientFactory, readerGroupManager, "RG" + RANDOM.nextInt(Integer.MAX_VALUE),
+                "R" + RANDOM.nextInt(Integer.MAX_VALUE));
+
+        readerGroupManager.close();
+        clientFactory.close();
+
+        controllerStarter.close(); // Shut down the controller
+
+        // Get names of all the segments created.
+        HashSet<String> allSegments = new HashSet<>(this.segmentStoreStarter.streamSegmentStoreWrapper.getSegments());
+        allSegments.addAll(this.segmentStoreStarter.tableStoreWrapper.getSegments());
+        log.info("No. of segments created = {}", allSegments.size());
+
+        // Get the long term storage from the running pravega instance
+        @Cleanup Storage tier2 = new AsyncStorageWrapper(new RollingStorage(this.storageFactory.createSyncStorage(),
+                new SegmentRollingPolicy(DEFAULT_ROLLING_SIZE)), DataRecoveryTestUtils.createExecutorService(1));
+
+        // wait for all segments to be flushed to the long term storage.
+        waitForSegmentsInStorage(allSegments, this.segmentStoreStarter.streamSegmentStoreWrapper, tier2)
+                .get(TIMEOUT.toMillis(), TimeUnit.MILLISECONDS);
+
+        this.segmentStoreStarter.close(); // Shutdown SegmentStore
+        this.segmentStoreStarter = null;
+        log.info("Segment Store Shutdown");
+
+        this.bkzk.close(); // Shutdown BookKeeper & ZooKeeper
+        this.bkzk = null;
+        log.info("BookKeeper & ZooKeeper shutdown");
+
+        // start a new BookKeeper and ZooKeeper.
+        this.bkzk = setUpNewBK(instanceId++);
+        this.dataLogFactory = new BookKeeperLogFactory(this.bkzk.bkConfig.get(), this.bkzk.zkClient.get(),
+                DataRecoveryTestUtils.createExecutorService(1));
+        this.dataLogFactory.initialize();
+
+        // Delete container metadata segment and attributes index segment corresponding to the container Id from the long term storage
+        DataRecoveryTestUtils.deleteContainerMetadataSegments(tier2, CONTAINER_ID);
+
+        // List all segments from the long term storage
+        Map<Integer, List<SegmentProperties>> segmentsToCreate = DataRecoveryTestUtils.listAllSegments(tier2, CONTAINER_COUNT);
+
+        // Start debug segment container using dataLogFactory from new BK instance and old long term storageFactory.
+        DebugTool debugTool = createDebugTool(this.dataLogFactory, this.storageFactory);
+        DebugStreamSegmentContainer debugStreamSegmentContainer = (DebugStreamSegmentContainer)
+                debugTool.containerFactory.createDebugStreamSegmentContainer(CONTAINER_ID);
+
+        // Re-create all segments which were listed.
+        Services.startAsync(debugStreamSegmentContainer, executorService)
+                .thenRun(new DataRecoveryTestUtils.Worker(debugStreamSegmentContainer, segmentsToCreate.get(CONTAINER_ID))).join();
+        sleep(5000); // Without sleep the test fails sometimes complaining some segment offsets don't exist.
```

Comment:
You need to figure out how to make this work without sleeping.

Reply:
Removed sleep. Doesn't fail on Travis, though fails locally sometimes. Will create an issue for it.

Reply:
It doesn't even fail locally now. Just added an extra line to wait for container metadata segment to be flushed to LTS, after debug segment container creates it in the Tier1.